### Reading external data 

In [1]:
import pandas as pd
import numpy as np

In [2]:
# open csv data - data needs to be in same directory or full absolute or relative path is needed!
df = pd.read_csv("benzin_ooe.csv")

In [3]:
# view excerpt - first few rows
df.head()

,Unnamed: 0;name;Super95;Diesel;PLZ;land;bundesland;bezirk;stadt;strasse;datum;coordinates_X;coordinates_Y
0,0;Lagerhaus Tankstelle;1.7990000000000002 EUR;...
1,1;Turmöl Quick;1.769 EUR;1.78 EUR;2091;at;Nied...
2,2;Genol Lagerhaus Tankstelle;0.0 EUR;1.7990000...
3,3;Turmöl Quick;1.714 EUR;1.734 EUR;4048;at;Obe...
4,4;AVIA Xpress (Automaten-Tankstelle);1.704 EUR...


-> depending on used locale (e.g. in DE/GER), excel could save CSV as "SSV" - semicolon separated values
also, e.g. German uses ',' as decimal point

In [6]:
# Pandas automatically adds index - if data has explicit index -> double index is not needed, so either drop first column afterwards or better - use:
df = pd.read_csv("benzin_ooe.csv", index_col=0, delimiter=';', decimal=',')

#if index has no name in csv, it is named "Unnamed:0" which is not nice - we can delete that name:
df.index.name = None

# show last few rows
df.tail()

,name,Super95,Diesel,PLZ,land,bundesland,bezirk,stadt,strasse,datum,coordinates_X,coordinates_Y
25664,Turmöl,0.0 EUR,1.898 EUR,4910,at,Oberösterreich,Ried,Ried im Innkreis,Braunauerstraße 10,29.09.2022 07:06,13.46058,48.20589
25665,TurboSprit,1.628 EUR,1.858 EUR,9130,at,Kärnten,Klagenfurt Land,Poggersdorf,Landesstrasse 2,29.09.2022 07:05,14.449627,46.641242
25666,Turmöl,1.614 EUR,1.874 EUR,1230,at,Wien,Liesing,Wien,Altmannsdorferstraße 150,29.09.2022 07:07,16.31935,48.154766
25667,Turmöl Quick,1.594 EUR,1.864 EUR,3150,at,Niederösterreich,St. Pölten Land,Wilhelmsburg,Lilienfelderstraße 37,29.09.2022 07:06,15.60052,48.10019
25668,Disk,1.689 EUR,1.883 EUR,6713,at,Vorarlberg,Bludenz,Ludesch,Walgaustraße 32,29.09.2022 07:06,9.77762,47.19298


In [8]:
# what datatypes do we have?
df.dtypes

name               str
Super95            str
Diesel             str
PLZ              int64
land               str
bundesland         str
bezirk             str
stadt              str
strasse            str
datum              str
coordinates_X      str
coordinates_Y      str
dtype: object

### 📝 Tasks for you:

In [7]:
# why so many strings? 
# 1) look at import function
pd.DataFrame({"is:":df.dtypes,"should be:":["str", "float64", "float64", "int64", "str", "str", "str", "str", "str","datetime64[us]", "float64", "float64"]})

,is:,should be:
name,str,str
Super95,str,float64
Diesel,str,float64
PLZ,int64,int64
land,str,str
bundesland,str,str
bezirk,str,str
stadt,str,str
strasse,str,str
datum,str,datetime64[us]


In [9]:
#What could you notice? Correctly import csv!

df = pd.read_csv("benzin_ooe.csv", index_col=0, delimiter=';', decimal=',')

#if index has no name in csv, it is named "Unnamed:0" which is not nice - we can delete that name:
df.index.name = None

In [10]:
pd.DataFrame({"is:":df.dtypes,"should be:":["str", "float64", "float64", "int64", "str", "str", "str", "str", "str","datetime64[us]", "float64", "float64"]})

,is:,should be:
name,str,str
Super95,str,float64
Diesel,str,float64
PLZ,int64,int64
land,str,str
bundesland,str,str
bezirk,str,str
stadt,str,str
strasse,str,str
datum,str,datetime64[us]


In [11]:
# still strings? look at csv - which ones got imported correctly? which not? how does the raw data of the wrong ones compare to the right ones?

,Super95,Diesel
0,1.7990000000000002 EUR,1.819 EUR
1,1.769 EUR,1.78 EUR
2,0.0 EUR,1.7990000000000002 EUR
3,1.714 EUR,1.734 EUR
4,1.704 EUR,1.712 EUR
5,1.779 EUR,0.0 EUR
6,1.789 EUR,1.7990000000000002 EUR
7,1.764 EUR,1.769 EUR
8,1.689 EUR,1.709 EUR
9,1.764 EUR,1.774 EUR


## Data types

In [13]:
#Columns are in German and names are inconsistent
df = df.rename( columns={"name": "Name", "PLZ": "ZIP", "land": "Country", "bundesland":"State",
                    "bezirk":"District", "stadt": "City", "strasse":"Address", "datum":"Date"} )

In [14]:
#ZIP-Code is read as a number, but we don't do calculations on it, it's better threated as an string so no statistics is calculated on it:
df.ZIP = df.ZIP.astype(str)

In [15]:
pd.DataFrame({"is:":df.dtypes,"should be:":["str", "float64", "float64", "str", "str", "str", "str", "str", "str","datetime65[us]", "float64", "float64"]})

,is:,should be:
Name,str,str
Super95,str,float64
Diesel,str,float64
ZIP,str,str
Country,str,str
State,str,str
District,str,str
City,str,str
Address,str,str
Date,str,datetime65[us]


In [16]:
# Fuel prices are still strings, let's have a look at the data
df.loc[:10,["Super95","Diesel"]]

,Super95,Diesel
0,1.7990000000000002 EUR,1.819 EUR
1,1.769 EUR,1.78 EUR
2,0.0 EUR,1.7990000000000002 EUR
3,1.714 EUR,1.734 EUR
4,1.704 EUR,1.712 EUR
5,1.779 EUR,0.0 EUR
6,1.789 EUR,1.7990000000000002 EUR
7,1.764 EUR,1.769 EUR
8,1.689 EUR,1.709 EUR
9,1.764 EUR,1.774 EUR


In [19]:
# remove 'EUR' -> replace ' EUR' with ''
# This is not working -> 'replace' works only for whole string, not parts
df[['Super95', 'Diesel']] = df[['Super95','Diesel']].replace(' EUR', '')

# Need to add 'regex=True' to replace substrings
df[['Super95', 'Diesel']] = df[['Super95', 'Diesel']].replace(' EUR', '', regex=True)

In [20]:
#what about those 1.7990000000000002? -> float precision errors, save to cut after 3 decimal digits (5 characters with 1.123)
df['Super95'] = df['Super95'].str[:5]
df['Diesel'] = df['Diesel'].str[:5]

In [21]:
# Fuel prices are still strings, let's have a look at the last 3 rows of fuel data
print(df.iloc[-3:][["Super95","Diesel"]], "\n\n", df[["Super95","Diesel"]].dtypes)

      Super95 Diesel
25666   1.614  1.874
25667   1.594  1.864
25668   1.689  1.883 

 Super95    str
Diesel     str
dtype: object


In [22]:
#after import, no automatic conversion, so need to explicitly be converted:
df[['Super95','Diesel']] = df[['Super95','Diesel']].astype(float)
df[['Super95','Diesel']].dtypes

Super95    float64
Diesel     float64
dtype: object

In [27]:
# Convert dates
df["Date"] = pd.to_datetime(df["Date"], dayfirst=True)
pd.DataFrame({"is:":df.dtypes,"should be:":["str", "float64", "float64", "str", "str", "str", "str", "str", "str","datetime64[us]", "float64", "float64"]})

,is:,should be:
Name,str,str
Super95,float64,float64
Diesel,float64,float64
ZIP,str,str
Country,str,str
State,str,str
District,str,str
City,str,str
Address,str,str
Date,datetime64[us],datetime64[us]


In [28]:
# what's with coordinates_X?
df.loc[[11732]]

,Name,Super95,Diesel,ZIP,Country,State,District,City,Address,Date,coordinates_X,coordinates_Y
11732,OIL! Tankstelle,1.789,0.0,6114,at,Tirol,Innsbruck Land,Kolsass,Auweg 1,2022-08-30 09:35:00,Nov.86,47.304


In [25]:
# Excel - oh! It's a date!

## Data cleaning

### Data capitalization

In [31]:
# Different capitalization in e.g. Cities / Addresses:
df.loc[[4, 5, 44, 106, 107, 300]]

,Name,Super95,Diesel,ZIP,Country,State,District,City,Address,Date,coordinates_X,coordinates_Y
4,Avia Xpress (Automaten-Tankstelle),1.704,1.712,4563,At,Oberösterreich,Kirchdorf,Micheldorf,Weinzierler Brücke 72,2022-08-13 17:57:00,14.129979,47.8940558
5,Avia Xpress (Unser Service: Luft Und Wasser),1.779,0.000,3754,At,Niederösterreich,Horn,Irnfritz 19,Irnfritz 19,2022-08-13 17:57:00,15.5578698,48.7399353
44,Shell Austria,0.000,1.709,9560,At,Kärnten,Feldkirchen,Feldkirchen,Dr.Arthur Lemischstrasse 16,2022-08-13 17:57:00,14.090068,46.726883
106,Shell Austria,1.739,1.739,1150,At,Wien,Rudolfsheim,Wien,Linzer Str 2A,2022-08-13 17:58:00,16.32173,48.19039
107,Shell Austria,0.000,1.719,5431,At,Salzburg,Hallein,Kuchl,Moos 66,2022-08-13 17:57:00,13.13875,47.6347
300,Inntaler Logistik Park Gmbh,1.698,1.768,6330,At,Tirol,Kufstein,Kufstein,Endach 34,2022-08-13 17:57:00,12.150235,47.565769


### Cleanup of Names

In [32]:
df["Name"] = df["Name"].str.title()
df["Country"] = df["Country"].str.title()
df["State"] = df["State"].str.title()
df["District"] = df["District"].str.title()
df["City"] = df["City"].str.title()
df["Address"] = df["Address"].str.title()

In [33]:
df.loc[[4, 5, 44, 106, 107, 300]]

,Name,Super95,Diesel,ZIP,Country,State,District,City,Address,Date,coordinates_X,coordinates_Y
4,Avia Xpress (Automaten-Tankstelle),1.704,1.712,4563,At,Oberösterreich,Kirchdorf,Micheldorf,Weinzierler Brücke 72,2022-08-13 17:57:00,14.129979,47.8940558
5,Avia Xpress (Unser Service: Luft Und Wasser),1.779,0.000,3754,At,Niederösterreich,Horn,Irnfritz 19,Irnfritz 19,2022-08-13 17:57:00,15.5578698,48.7399353
44,Shell Austria,0.000,1.709,9560,At,Kärnten,Feldkirchen,Feldkirchen,Dr.Arthur Lemischstrasse 16,2022-08-13 17:57:00,14.090068,46.726883
106,Shell Austria,1.739,1.739,1150,At,Wien,Rudolfsheim,Wien,Linzer Str 2A,2022-08-13 17:58:00,16.32173,48.19039
107,Shell Austria,0.000,1.719,5431,At,Salzburg,Hallein,Kuchl,Moos 66,2022-08-13 17:57:00,13.13875,47.6347
300,Inntaler Logistik Park Gmbh,1.698,1.768,6330,At,Tirol,Kufstein,Kufstein,Endach 34,2022-08-13 17:57:00,12.150235,47.565769


In [34]:
for col in df.columns:
    if df[col].dtype == "str":
        df[col] = df[col].replace(" An Der ", " a. d. ", regex=True)
        df[col] = df[col].replace(" A. D. ", " a. d. ", regex=True)
        df[col] = df[col].replace(" Im ", " im ", regex=True)
        df[col] = df[col].replace(" In ", " in ", regex=True)
        df[col] = df[col].replace(" Bei ", " bei ", regex=True)
        df[col] = df[col].replace(" B. ", " bei ", regex=True)
        df[col] = df[col].replace("Sankt ", "St.", regex=True)
        df[col] = df[col].replace(r"St\. ", "St.", regex=True)
        df[col] = df[col].replace("Wr. Neu", "Wiener Neu", regex=True)
        df[col] = df[col].replace("Puch/Hallein", "Puch bei Hallein")
        df[col] = df[col].replace("Puch-Hallein", "Puch bei Hallein")
        df[col] = df[col].replace("Waidhofen/Thaya", "Waidhofen a. d. Thaya")
        df[col] = df[col].replace("Bruck / Leitha", "Bruck a. d. Leitha")
        df[col] = df[col].replace("Bruck/Leitha", "Bruck a. d. Leitha")
        df[col] = df[col].replace("Brueckl", "Brückl")
        df[col] = df[col].replace("D-Ehrensdorf", "Deutsch Ehrensdorf")
        df[col] = df[col].replace("Deutsch-Griffen", "Deutsch Griffen")
        df[col] = df[col].replace("Deutsch-Wagram" , "Deutsch Wagram")
        
        


In [35]:
df["Name"] = df["Name"].replace("Bp", "BP")
df["Name"] = df["Name"].replace("Diskont-Tankstelle", "Diskont Tankstelle", regex=True)


In [36]:
df["Address"] = df["Address"].replace("Am Hofer Parkplatz", "Hofer Pp.", regex=True)
df["Address"] = df["Address"].replace("Hofer Parkplatz", "Hofer Pp.", regex=True)
df["Address"] = df["Address"].replace("Hofer-Parkplatz", "Hofer Pp.", regex=True)

In [37]:
df["Address"] = df["Address"].replace(r"str\.", "straße", regex=True)
df["Address"] = df["Address"].replace(r"Str\.", "Straße", regex=True)
df["Address"] = df["Address"].replace("trasse", "traße", regex=True)
df["Address"] = df["Address"].replace("staße", "straße", regex=True)
df["Address"] = df["Address"].replace("traße;", "traße", regex=True)
df["Address"] = df["Address"].replace("Vis A Vis", "vav.", regex=True)

## Gaining Data Overview

In [38]:
# View some basic statistics of string columns
df[["ZIP", "Country", "State", "District", "City", "Address"]].describe()

,ZIP,Country,State,District,City,Address
count,25669,25669,25669,25669,25669,25667
unique,986,1,9,96,1123,1998
top,9020,At,Niederösterreich,Salzburg Umgebung,Wien,Bahnhofstraße 26
freq,297,25669,6827,574,2446,83


In [42]:
# it's all from Austria... so no info there.
df = df.drop(columns=["Country"])
df.columns

KeyError: "['Country'] not found in axis"

### 📝 Tasks for you:

In [43]:
# Answer the following questions by assigning correct values to the variables:

# how many entries are in the dataset?
nr_of_rows = -1
# how many different states are in the dataset?
nr_of_states = -1
# how many different cities are in the dataset?
nr_of_cities = -1

# "Country" only has 1 unique entry: "at" (all the columns are the same), which doesn't provide any useful info - delete it!
df = df

In [44]:
# View some basic statistics of number columns
df.describe()

,Super95,Diesel,Date
count,25669.000000,25669.000000,25669
mean,1.243556,1.500192,2022-08-27 17:18:06.460711
min,0.000000,0.000000,2022-02-09 06:13:00
25%,0.000000,1.714000,2022-08-20 08:37:00
50%,1.686000,1.818000,2022-08-29 07:02:00
75%,1.729000,1.899000,2022-09-17 07:29:00
max,16.630000,1775.000000,2022-12-09 01:59:00
std,0.772618,11.097348,NaN


In [45]:
# calculate the mean price of Super and Diesel
df[["Super95", "Diesel"]].mean()

Super95    1.243556
Diesel     1.500192
dtype: float64

### Do those prices seem ok to you? What were the mean prices of Diesel or Super95 in 2022?

2022: one of the most expensive year for petrol cars: 
https://www.adac.de/verkehr/tanken-kraftstoff-antrieb/deutschland/kraftstoffpreisentwicklung/#teuerstes-tankjahr-2022

Prices (in GER) about
- Super95    1.86
- Diesel     1.94

Was fuel in Austria so much cheaper?

Why is mean and median so far away?<br>Why is standard deviation so big?

### 📝 Tasks for you:

In [ ]:
# What do you think is the problem with the calculation of the average prices? Write your answer between the triple quotes:
avg_calc_answer = """

"""

In [47]:
# if you plan to stay in python with the data, to save the cleaned dataframe, just use:
df.to_csv("fuel_prices_cleaned.csv")

# if you want to edit it in Microsoft Excel, without import wizard,
# you have to add 'encoding="utf-8-sig"' or it will show 'trash' characters.
# this might be different for your operating system, tools, language and locale settings etc.
# ⚠️ when saving, delimiter is called separator (sep)

# to save the file compatible with German locale settings, compatible with Excel in Windows:
# df.to_csv("fuel_prices.csv", encoding="utf-8-sig", sep=';', decimal=',')

# ⚠️ if you edit a csv in Excel, you have to export it as CSV UTF-8 to correctly load in pyton! (there are several CSV export options!)